### ISSUE TITLE
Credit Card Tokenizer Issue

### ISSUE THREAD

**Issue by engjellselaci** [Charity-Shop-Exchange/Charity-Shop-Exchange#162]
Library used to create the UI: React, Gatsby

The library that we are importing from billsby to create tokenizers is not working for some reason. They do not have github nor npm or any kind of forum that people talk about it.
This is the error im getting: 
![image](https://user-images.githubusercontent.com/47112077/88279489-d3717e00-ccec-11ea-933e-3969c13662ff.png)

Let me know if I should explain the process of how I'm trying to generate credit card token.
Thanks.

**Comment by engjellselaci** [comment]
Here's where I'm trying to tokenize my credit card details:
`import React, { useState, useEffect, useCallback } from 'react';
import ModalButton from './ModalButton';
import Logo from '../../../assets/img/cse_logo.png';
import billsbyConfig from '../../../import/billsby';

export default function Footer(props) {
  const { formData, handleChange } = props;
  let [newClass, setClassName] = useState('');
  let [button, changeButton] = useState('modal-button');

  const [quantity, setQuantity] = useState(true);
  const [isReady, setIsReady] = useState(false);

  useEffect(() => {
      window.billsbyTokens.on('ready', function() {
        setIsReady(true);
      });
      window.billsbyTokens.on("paymentMethod", function (token, pmData) {

        console.log('make axios here');
      });
  }, []);
  const handleSubmit = useCallback(() => {
    const requiredFields = {
      full_name: formData.payment.name,
      month: formData.payment.expiry_at.split('/')[0],
      year: formData.payment.expiry_at.split('/')[1],
    };


    window.billsbyTokens.tokenizeCreditCard(requiredFields);

    setTimeout(() => {
      console.log('Trigger scan');
      window.scanDomBillsby();
      setTimeout(() => {
        const elem = document.getElementById('billsbyTriggerAnchor');
        console.log(elem);
        if (elem) {
          console.log('Click elem');
          elem.click();
        }
      }, 500);
    }, 500);
  }, [formData]);
  const books = formData.checkoutItems.books.filter(
    (item) => item.quantity !== 0,
  );
  const dvd = formData.checkoutItems.dvd.filter((item) => item.quantity !== 0);
  const total = books.concat(dvd);
  const orderSummary = total.reduce((quantity, item) => {
    for (const [orderName, orderCount] of Object.entries(item)) {
      if (!quantity[orderName]) {
        quantity[orderName] = 0;
      }

      quantity[orderName] += orderCount;
    }
    return quantity;
  }, {});

  const totalSum = total.reduce((sum, i) => (sum += i.quantity * 2.0), 3.5);
  const totalDecimalSum = (Math.round(totalSum * 100) / 100).toFixed(2);

  const keysToLook = [
    'address_1',
    'firstname',
    'lastname',
    'postcode',
    'town',
    'email',
  ];

  const showToolTip = keysToLook.some((key) => formData.delivery[key] === '');

  return (
    <div className="modal-footer" id="footer">
      <img
        className="h-10 md:h-12 img-logo"
        src={Logo}
        alt="Charity Shop Exchange"
      />
      <Selection
        orderSummary={orderSummary}
        formData={formData}
        handleSubmit={handleSubmit}
        handleChange={handleChange}
      />
      <Subscription
        text="delivered every"
        formData={formData}
        button={button}
        showToolTip={showToolTip}
        /* showToolTip2={showToolTip2} */
        quantity={quantity}
        orderSummary={orderSummary}
        handleSubmit={handleSubmit}
        handleChange={handleChange}
        newClass={newClass}
        totalDecimalSum={totalDecimalSum}
        isReady={isReady}
      />
      <a
        style={{ display: 'none' }}
        id="billsbyTriggerAnchor"
        data-billsby-type="checkout"
        data-billsby-units={orderSummary.quantity}>
        Subscribe
      </a>
    </div>
  );
}
const Selection = ({ orderSummary, handleSubmit, handleChange }) => {
  if (orderSummary.quantity >= 2) {
    return null;
  }
  if (orderSummary.quantity >= 1) {
    return (
      <>
        <ModalButton
          className="modal-button-disabled"
          disabled={
            typeof orderSummary.quantity !== 'undefined'
              ? orderSummary.quantity < 2
              : true
          }
          onClick={handleSubmit}
          handleChange={handleChange}
          text="Select 1 more item"
        />
      </>
    );
  }
  return (
    <>
      <ModalButton
        className="modal-button-disabled-2"
        disabled={
          typeof orderSummary.quantity !== 'undefined'
            ? orderSummary.quantity < 2
            : true
        }
        onClick={handleSubmit}
        handleChange={handleChange}
        text="Select 2 more items"
      />
    </>
  );
};
const Subscription = ({
  formData,
  orderSummary,
  newClass,
  button,
  handleSubmit,
  quantity,
  handleChange,
  showToolTip,
  showToolTip2,
  totalDecimalSum,
  isReady
}) => {
  const modalButtonClassName = `${
    showToolTip2 || showToolTip ? 'modal-button-disabled' : button
    } `;
  const isEnabled = !showToolTip && !showToolTip2;
  if (
    typeof orderSummary.quantity !== 'undefined'
      ? orderSummary.quantity < 2
      : true
  ) {
    return null;
  }
  return (
    <div className="w-full flex justify-center flex-wrap">
      {quantity ? (
        <span className="p text-lg justify-center text-center flex-grow">
          <a className="s">{orderSummary.quantity} items</a>delivered every
          <a className="s2">{formData.delivery.subscription}</a>
        </span>
      ) : (
          <div className="fade-in">
            <span>Thank You! An email confirmation has been sent.</span>
          </div>
        )}
      <ModalButton
        className={modalButtonClassName}
        type="button"
        id="submit-button"
        disabled={!isEnabled && !isReady}
        onClick={handleSubmit}
        handleChange={handleChange}>
        <>
          {quantity ? (
            <span className="sum">£{totalDecimalSum} Subscribe</span>
          ) : (
              <span className="subscribe">Subscribed</span>
            )}
          {/* {showToolTip2 ? (
            <span className="tooltiptext">Enter Payment Details</span>
          ) : (
            ''
          )} */}
          {showToolTip ? (
            <span className="tooltiptext">Enter Delivery Details</span>
          ) : (
              ''
            )}
          <div className={newClass}></div>
        </>
      </ModalButton>
    </div>
  );
};
`

**Comment by kelvinlockwood** [comment]
slack thread here... https://charityshopexchange.slack.com/archives/C0106LUBN7K/p1595441543008800

**Comment by engjellselaci** [comment]
I have initialized the billsby library , but the problem still shows on devtools. This is the code i am using:

`  useEffect(() => {

    try {
      window.billsbyTokens.init("billsby-number", "billsby-cvv");
    } catch (err) {
      console.error(err);
    }
    window.billsbyTokens.on('ready', function () {
      setIsReady(true);
    });
    window.billsbyTokens.on("paymentMethod", function (token, pmData) {

      console.log('make axios here');
    });
  }, []);

  const handleSubmit = useCallback(() => {
    const requiredFields = {
      full_name: payment.name,
      month: payment.expiry_at.split('/')[0],
      year: payment.expiry_at.split('/')[1],
    };
    try {
      window.billsbyTokens.tokenizeCreditCard(requiredFields);
    } catch (err) {
      console.log(err)
    }
})
`
The error still appears:
![Screenshot from 2020-07-30 21-04-29](https://user-images.githubusercontent.com/47112077/88957908-53c54f80-d2a8-11ea-82bd-a97a532dbef9.png)

We should contact billsby again and see if they can provide us with any solution to this , any help is appreciated.

**Comment by hmajid2301** [comment]
let's try to initialize it with actual values like instead of `billsby-number` use the actual div.

**Comment by kelvinlockwood** [comment]
_Message from Alex at Billsby:_

Hello guys,

I checked the snippet on github, it's not the full code so i can just guess what is the actual issue, make sure that:

- "billsby-number" and "billsby-cvv" arguments passed to the billsbyTokens.init function are the actual ID's of the two html elements in the DOM.

- debug the code and make sure that the call billsbyTokens.tokenizeCard triggers after you call the billsbyTokens.init

handleSubmit is a memoized callback but i can't see when it's called in your code, it might be called before the billsbyTokens.init and before you register the billsbyTokens.on listeners, that would trigger the error you're seeing in the console.
debug the code and make sure the sequential order of the calls is the one mentioned above.

**Comment by sajjadatma** [comment]
hay guys, i I checked the your codes on github and I think this should not be all your codes. 
when you add billsby on your project (as stated in the document) billsby will create a iframe tag for tokenizers and this iframe tag is your issue. 
for the first step :
bringing out this 2 line from useEffect,
`     window.billsbyTokens.on('ready', function() {
        setIsReady(true);`
and 
step 2:
check this link for iframe issue 
[https://davidwalsh.name/iframe-contentwindow-null](url) 
I think by this two step ,your problem will solve.

**Comment by engjellselaci** [comment]
```

import React, { useState, useEffect, useLayoutEffect } from 'react';
import Collapsable from '../Collapsable';
import TextFieldWithLabel from '../../TextFieldWithLabel';
import Visa from '../../../assets/img/icons8-visa.svg';
import MasterCard from '../../../assets/img/icons8-mastercard.svg';
import Discover from '../../../assets/img/icons8-discover.svg';
import Card1 from '../../../assets/img/card1.png';
import Card2 from '../../../assets/img/card2.png';
import Card3 from '../../../assets/img/card3.png';
import EmailMeWhenSubscribed from './EmailMeWhenSubscribed';

import useFormDataValidation, {
  isRequiredPayment,
  validNumber,
  isValidCardName,
  isValidEmail,
  validCreditCard,
  validCvv,
  validDate,
  phoneNumber,
} from '../../../hooks/useFormDataValidation';

const valid = require('card-validator');

export default function Payment(props) {
  const { formData, handleChange } = props;
  const { payment, delivery } = formData;
  const [isOpen, setisOpen] = useState(false);
  
  useEffect(() => {
  
    try {
      window.billsbyTokens.init("billsby-number", "billsby-cvv");
  } catch (err) {
      console.error(err);
    }  
    window.billsbyTokens.on('ready',  () => {
      console.log('Event Happened');

    });
    window.billsbyTokens.on("paymentMethod", function (token, pmData) {

      console.log('make axios here');
    });
  }, []);

  const fieldValidators = {
    phone: [isRequiredPayment, validNumber],
    email: [isRequiredPayment, isValidEmail],
    card_number: [isRequiredPayment, validCreditCard],
    expiry_at: [isRequiredPayment, validDate],
    cvv: [isRequiredPayment, validCvv],
    name: [isRequiredPayment, isValidCardName],
  };
  const [initErrors, setInitErrors] = useState({});

  const formSubmitAction = (values) => {
    console.log(values);
  };

  const { values, errors, fieldChange } = useFormDataValidation(
    formData,
    initErrors,
    formSubmitAction,
    fieldValidators,
  );

  const handleSubmit = () => {
    console.log('Submit', formData);
  };

  return (
    <Collapsable
      title="Payment"
      open={isOpen}
      toggle={() => setisOpen(!isOpen)}>
      <form id="payment-form" action="https://www.billsby.com/">
        
        <div className="flex flex-col">
          <div className="payment-textbox-inner-width" autoCorrect="off">
            <div className="relative">
              <TextFieldWithLabel
                autocomplete="cc-number"
                type="cc-number"
                label="Card details"
                htmlFor="cc-number"
                placeholder="1234 1234 1234 1234"
                name="cc-number"
                pattern="[0-9\s]{13,19}"
                max={16}
                id="billsby-number"
                keyToUpdate="payment.card_number"
                value={payment.card_number}
                onChange={(value) => {
                  handleChange(value);
                }}
                required={true}
                onblur={(event) => fieldChange(event, 'card_number')}
              />
              <Card
                valid={valid}
                formData={formData}
                Card1={Card1}
                Card2={Card2}
                Card3={Card3}
                Discover={Discover}
                MasterCard={MasterCard}
                Visa={Visa}
              />
            </div>
            {/* <p style={{fontSize: '14px', padding: '0 10px', color: '#c53030'}}>{errors.card_number}</p> */}
          </div>
        </div>
        <div className="payment-textbox-inner-width">
          <div style={{ display: 'flex' }}>
            <div className="relative" style={{ width: '50%' }}>
              <TextFieldWithLabel
                autocomplete="cc-exp"
                type="cc-exp"
                placeholder="MM/YY"
                name="cc-exp"
                mask={'00/0000'}
                keyToUpdate="payment.expiry_at"
                // className="w-6/12"
                title="MM/YY"
                value={payment.expiry_at}
                pattern="(1[0-2]|0[1-9])\/(1[5-9]|2\d)"
                s
                max={5}
                required={true}
                onChange={(value) => {
                  handleChange(value);
                }}
                onblur={(event) => fieldChange(event, 'expiry_at')}
              />
            </div>
            <div className="relative" style={{ width: '50%' }}>
              <TextFieldWithLabel
                autocomplete="cc-csc"
                placeholder="CVV"
                type="cc-csc"
                id="billsby-cvv"
                name="cc-csc"
                keyToUpdate="payment.cvv"
                // className="w-6/12"
                title="CVV"
                value={payment.cvv}
                required={true}
                onChange={(value) => {
                  handleChange(value);
                }}
                max={4}
                onblur={(event) => fieldChange(event, 'cvv')}
              />
              <CVVIMG
                valid={valid}
                formData={formData}
                Card1={Card1}
                Card2={Card2}
                Card3={Card3}
                Discover={Discover}
                MasterCard={MasterCard}
                Visa={Visa}
              />
            </div>
          </div>
          <div style={{ display: 'flex' }}>
            {/* <p style={{fontSize: '14px', padding: '0 10px', color: '#c53030'}}>{errors.expiry_at}</p>
          <p style={{fontSize: '14px', padding: '0 16px', color: '#c53030'}}>{errors.cvv}</p> */}
          </div>
        </div>
        <div className="flex flex-col">
          <div className="payment-textbox-inner-width">
            <div className="relative">
              <TextFieldWithLabel
                autocomplete="cc-name"
                type="cc-name"
                label="Name on card"
                placeholder=""
                required={false}
                name="cc-name"
                id="full_name"
                keyToUpdate="payment.name"
                value={payment.name}
                onChange={(value) => {
                  handleChange(value);
                }}
                onBlur={(event) => fieldChange(event, 'name')}
              />
              {/* <p className="error">{errors.name}</p> */}
            </div>
          </div>
        </div>
        {/* <div className="mt-4">
        <div className="delivery-frequency">Your's:</div>
      </div> */}
        <div className="flex flex-col">
          <div className="payment-textbox-inner-width">
            <div className="relative">
              <TextFieldWithLabel
                autocomplete="number"
                label="Phone number"
                name="number"
                type="number"
                keyToUpdate="payment.phone"
                value={payment.phone}
                onChange={(value) => {
                  handleChange(value);
                }}
                title="May be used to assist delivery"
                onblur={(event) => fieldChange(event, 'phone')}
              />
              <p className="error">
                {payment.phone &&
                  payment.phone.length > 0 &&
                  validNumber(payment.phone)}
              </p>
            </div>
          </div>
          <div className="payment-textbox-inner-width ">
            <div className="relative">
              <TextFieldWithLabel
                autocomplete="email"
                label="Email"
                type="email"
                name="email"
                keyToUpdate="payment.email"
                value={payment.email}
                pattern="^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$"
                required={true}
                error={
                  payment.email &&
                  payment.email.length > 0 &&
                  isValidEmail(payment.email)
                }
                onChange={(value) => {
                  handleChange(value);
                }}
                onblur={(event) => fieldChange(event, 'email')}
              />
            </div>
          </div>
          <div
            className="mt-4"
            style={{ color: '#696969', padding: '0 8px', margin: 0 }}>
            <EmailMeWhenSubscribed
              keyToUpdate="payment.isEmailedMe"
              onChange={(value) => {
                handleChange(value);
              }}
              lable="Yes, email me about special offers and new product information"
            />
          </div>
        </div>
        {!delivery.is_used_same_address_for_billing ? (
          <>
            <div className="mt-4">
              <div className="delivery-frequency">Billing address:</div>
            </div>
            <div className="flex flex-col flex-wrap">
              {/* <div className="md:w-1/2 pr-1">
              <div className="relative">
                <TextFieldWithLabel
                  label="Full name"
                  name="payment.billing_fullname"
                  value={payment.billing_fullname}
                  onChange={handleChange}
                />
              </div>
            </div> */}
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete="name"
                    label={'First name'}
                    name="name"
                    type="name"
                    keyToUpdate="payment.billing_firstname"
                    value={payment.billing_firstname}
                    pattern=".{2,}"
                    required={true}
                    error={
                      values.billing_firstname &&
                      values.billing_firstname.length > 0 &&
                      !/^.{2,}$/.test(values.billing_firstname) &&
                      'Please enter recipients first name'
                    }
                    onChange={handleChange}
                    onblur={(event) => fieldChange(event, 'billing_firstname')}
                  />
                </div>
              </div>
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete="name"
                    label={'Last name'}
                    type="name"
                    name="name"
                    keyToUpdate="payment.billing_lastname"
                    value={payment.billing_lastname}
                    pattern=".{2,}"
                    required={true}
                    error={
                      values.billing_lastname &&
                      values.billing_lastname.length > 0 &&
                      !/^.{2,}$/.test(values.billing_lastname) &&
                      'Please enter recipients last name'
                    }
                    onChange={handleChange}
                    onblur={(event) => fieldChange(event, 'billing_lastname')}
                  />
                </div>
              </div>
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete="postal-code"
                    label="UK Postcode"
                    name="postal-code"
                    type="postal-code"
                    max={8}
                    keyToUpdate="payment.billing_postcode"
                    required={true}
                    value={payment.billing_postcode}
                    pattern="^[A-Z]{1,2}\d[A-Z\d]? ?\d[A-Z]{2}$"
                    error={
                      values.billing_postcode &&
                      values.billing_postcode.length > 0 &&
                      !/^[A-Z]{1,2}\d[A-Z\d]? ?\d[A-Z]{2}$/.test(
                        values.billing_postcode,
                      ) &&
                      'Please enter a valid postcode'
                    }
                    onblur={(event) => fieldChange(event, 'billing_postcode')}
                    onChange={handleChange}
                  />
                </div>
              </div>
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete="address-line1"
                    label="Address line 1"
                    name="address-line1"
                    type="address-line1"
                    keyToUpdate="payment.billing_address_1"
                    required={true}
                    max={255}
                    pattern=".{1,}"
                    error={
                      values.billing_address_1 &&
                      values.billing_address_1.length > 0 &&
                      !/^.{1,}$/.test(values.billing_address_1) &&
                      'Please enter a valid address'
                    }
                    value={payment.billing_address_1}
                    onblur={(event) => fieldChange(event, 'billing_address_1')}
                    onChange={handleChange}
                  />
                </div>
              </div>
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete="address-line2"
                    label="Address line 2"
                    type="address-line2"
                    name="address-line2"
                    max={255}
                    keyToUpdate="payment.billing_address_2"
                    value={payment.billing_address_2}
                    onblur={(event) => fieldChange(event, 'billing_address_2')}
                    onChange={handleChange}
                  />
                </div>
              </div>
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete="address-level1"
                    label="Town / City"
                    name="address-level1"
                    type="address-level1"
                    keyToUpdate="payment.billing_town"
                    max={35}
                    pattern=".{2,}"
                    required={true}
                    error={
                      values.billing_town &&
                      values.billing_town.length > 0 &&
                      !/^.{2,}$/.test(values.billing_town) &&
                      'Please enter the postal town or city'
                    }
                    value={payment.billing_town}
                    onblur={(event) => fieldChange(event, 'billing_town')}
                    onChange={handleChange}
                  />
                </div>
              </div>
              <div className="md:w-1/2 pr-1">
                <div className="relative">
                  <TextFieldWithLabel
                    autocomplete=""
                    type="county"
                    label="County"
                    max={35}
                    required={true}
                    keyToUpdate="payment.billing_county"
                    value={payment.billing_county}
                    onblur={(event) => fieldChange(event, 'billing_county')}
                    onChange={handleChange}
                  />
                </div>
              </div>
            </div>
          </>
        ) : null}
      </form>
      {/* <div className="mt-4">
      <Button theme="small" onClick={handleSubmit}>
        Place Order
      </Button>
    </div> */}

    </Collapsable>
  );
}
const Card = ({
  formData,
  valid,
  Visa,
  MasterCard,
  Discover,
  Card1,
  Card2,
  Card3,
}) => {
  const cardValidator = valid.number(formData.payment['card_number'], {
    luhnValidateVisaCard: false,
  });
  if (cardValidator.card) {
    return (
      <div className="payment-card-img">
        <img
          className=""
          src={
            cardValidator.card &&
              cardValidator.card.type &&
              cardValidator.card.type === 'mastercard'
              ? MasterCard
              : Card1
          }
          alt="card1"
        />
        <img
          className=""
          src={
            cardValidator.card &&
              cardValidator.card.type &&
              cardValidator.card.type === 'discover'
              ? Discover
              : Card2
          }
          alt="card2"
        />
        <img
          className=""
          src={
            cardValidator.card &&
              cardValidator.card.type &&
              cardValidator.card.type === 'visa'
              ? Visa
              : Card3
          }
          alt="card3"
        />
      </div>
    );
  }
  return (
    <>
      <div className="payment-card-img">
        <img className="" src={Card1} alt="card1" />
        <img className="" src={Card2} alt="card2" />
        <img className="" src={Card3} alt="card3" />
      </div>
    </>
  );
};
const CVVIMG = ({
  formData,
  valid,
  Visa,
  MasterCard,
  Discover,
  Card1,
  Card2,
  Card3,
}) => {
  const cardValidator = valid.number(formData.payment['card_number'], {
    luhnValidateVisaCard: false,
  });
  if (cardValidator.card) {
    return (
      <div className="payment-card-img">
        <img
          className=""
          src={
            cardValidator.card &&
              cardValidator.card.type &&
              cardValidator.card.type === 'mastercard'
              ? MasterCard
              : cardValidator.card &&
                cardValidator.card.type &&
                cardValidator.card.type === 'visa'
                ? Visa
                : cardValidator.card &&
                  cardValidator.card.type &&
                  cardValidator.card.type === 'discover'
                  ? Discover
                  : Card1
          }
          alt="cvv"
        />
      </div>
    );
  }
  return (
    <div className="payment-card-img">
      <img className="" src={Card1} alt="card1" />
    </div>
  );
};

```

I am initializing billsby after mounting the component, but im still getting the error. 
Feel free to jump in the code and let me know what im doing wrong.

**Comment by sajjadatma** [comment]
![88279489-d3717e00-ccec-11ea-933e-3969c13662ff](https://user-images.githubusercontent.com/25636184/89244619-2ec03c00-d61c-11ea-9dea-5c0fd3425f56.png)

See this image.
Usually Js files initializing before body. So as you see in this image Iframe-vm.min.js initializing as first js file , even before bootatrap. I think i frame should be initiated after body.
I think if you use onload() on script tag in html.js you issue will solve.

**Comment by hmajid2301** [comment]
I've solved the above issue it was indeed related to id's not being present. I have encountered another issue after sending the tokenizer request:

`Object { key: "errors.invalid", message: "Credit card number entered is not valid", attribute: "number" }`

I wanted to ask what format does it expect the card number and cvv to be in? Currently they are text inputs. I noticed in the example you are using divs. I tried divs but this also returned the same error. Is there way to check the data being sent to billsby thanks.
